include/zephyr/bindesc.h

```diff
@@ -0,0 +1,405 @@
+
+/** The day of the month the image was compiled in */
+#define BINDESC_ID_BUILD_TIME_DAY 0xa02
+
+/** The hour of the day the image was compiled in */
+#define BINDESC_ID_BUILD_TIME_HOUR 0xa03
+
+/** The minute the image was compiled in */
+#define BINDESC_ID_BUILD_TIME_MINUTE 0xa04
+
+/** The second the image was compiled in */
+#define BINDESC_ID_BUILD_TIME_SECOND 0xa05
+
+/** The UNIX time (seconds since midnight of 1970/01/01) the image was compiled in */
+#define BINDESC_ID_BUILD_TIME_UNIX 0xa06
+
+/** The date and time of compilation such as "2023/02/05 00:07:04" */
+#define BINDESC_ID_BUILD_DATE_TIME_STRING 0xa07
+
+/** The date of compilation such as "2023/02/05" */
+#define BINDESC_ID_BUILD_DATE_STRING 0xa08
+
+/** The time of compilation such as "00:07:04" */
+#define BINDESC_ID_BUILD_TIME_STRING 0xa09
+
+/** The name of the host that compiled the image */
+#define BINDESC_ID_HOST_NAME 0xb00
+
+/** The C compiler name */
+#define BINDESC_ID_C_COMPILER_NAME 0xb01
+
+/** The C compiler version */
+#define BINDESC_ID_C_COMPILER_VERSION 0xb02
+
+/** The C++ compiler name */
+#define BINDESC_ID_CXX_COMPILER_NAME 0xb03
+
+/** The C++ compiler version */
+#define BINDESC_ID_CXX_COMPILER_VERSION 0xb04
+
+#define BINDESC_TAG_DESCRIPTORS_END BINDESC_TAG(DESCRIPTORS_END, 0x0fff)
+
+/**
+ * @cond INTERNAL_HIDDEN
+ */
+
+/*
+ * Utility macro to generate a tag from a type and an ID
+ *
+ * type - Type of the descriptor, UINT, STR or BYTES
+ * id - Unique ID for the descriptor, must fit in 12 bits
+ */
+#define BINDESC_TAG(type, id) ((BINDESC_TYPE_##type & 0xf) << 12 | (id & 0x0fff))
+
+/*
+ * Utility macro to get the name of a bindesc entry
+ */
+#define BINDESC_NAME(name) bindesc_entry_##name
+
+/*
+ * Utility macro to convert a uint32_t to a uint8_t array
+ */
+#if IS_ENABLED(CONFIG_LITTLE_ENDIAN)
+#define UINT_TO_BYTE_ARRAY(num) {		\
+	((num) & 0xff),				\
+	(((num) >> 8) & 0xff),			\
+	(((num) >> 16) & 0xff),			\
+	(((num) >> 24) & 0xff)}
+#else
+#define UINT_TO_BYTE_ARRAY(num) {		\
+	(((num) >> 24) & 0xff),			\
+	(((num) >> 16) & 0xff),			\
+	(((num) >> 8) & 0xff),			\
+	((num) & 0xff)}
+#endif /* IS_ENABLED(CONFIG_LITTLE_ENDIAN) */
+
+/* Convenience helper for declaring a binary descriptor entry. */
+#define DEFINE_BINDESC_ENTRY(name)					\
+	__aligned(BINDESC_ALIGNMENT) const struct bindesc_entry BINDESC_NAME(name)	\
+	__in_section(_bindesc_entry, static, name) __used __noasan
+
+/**
+ * @endcond
+ */
+
+/**
+ * @brief Define a binary descriptor of type string.
+ *
+ * @details
+ * Define a string that is registered in the binary descriptor header.
+ * The defined string can be accessed using @ref BINDESC_GET_STR
+ *
+ * @note The defined string is not static, so its name must not collide with
+ * any other symbol in the executable.
+ *
+ * @param name Name of the descriptor
+ * @param id Unique ID of the descriptor
+ * @param value A string value for the descriptor
+ */
+#define DEFINE_BINDESC_STR(name, id, value)	\
```

Comment:
Not a fan of the `DEFINE_` prefix. Not blocking, but could we drop it?

Reply:
Originally this was to be consistent with the rest of Zephyr's API, but now I realize that the rest of the API uses it as a postfix (e.g. K_THREAD_DEFINE, K_TIMER_DEFINE, K_QUEUE_DEFINE etc.). So I can either remove it or change is to a postfix, whatever you decide.

Reply:
IMHO the <SUBSYS>_<OP> naming convention should be used. It is also more convenient when using auto completion because you type several letters of subsys and then select among provided defines only for the subsys.

Reply:
> IMHO the _ naming convention should be used. It is also more convenient when using auto completion because you type several letters of subsys and then select among provided defines only for the subsys.

@de-nordic What's the _ naming convention? Do you mean that everything should start with `BINDESC_`?

Reply:
@yonsch let's switch to `_DEFINE` as a postfix.

Reply:
This is the last blocking bit from my side.

Reply:
> > IMHO the _ naming convention should be used. It is also more convenient when using auto completion because you type several letters of subsys and then select among provided defines only for the subsys.
> 
> @de-nordic What's the _ naming convention? Do you mean that everything should start with `BINDESC_`?

Yes, I think so.